### PR TITLE
Quake/manager

### PR DESCRIPTION
**Pull request by athena255**
The Quake Manager

### CODE REVIEW

File: test/test_quake/quake_common.h [hunk 1]
```diff
@@ -18,14 +19,14 @@ typedef struct sbdix_r {
...
+const int MT_NO_MSG     = 0;    // No SBD message received from GSS
+const int MT_MSG_RECV   = 1;    // SBD message successfully retrieved
+const int MT_MSG_ERR    = 2;    // Error occured
 
 // MO response statuses
-const int MO_OK = 0; // msg transfer ok
-const int MO_TOO_BIG = 1; // transfer ok but msg in queue too big for transfer
-const int MO_NO_UPDATE = 2; // transfer ok but no location update
+const int MO_OK         = 0;    // msg transfer ok
+const int MO_TOO_BIG    = 1;    // transfer ok but msg in queue too big for transfer
```
Comment (tanishqaggarwal) [inline test/test_quake/quake_common.h]: Maybe clarify by saying msg in the _MT_ queue is too big for transfer?

File: test/test_quake/quake_common.h [hunk 2]
```diff
@@ -34,3 +35,45 @@ void setUp(void) {
 
 void tearDown(void) {
 }
+
+// length of a ctrl cycle in ms
+static const int DEFAULT_CTRL_CYCLE_LENGTH = 140;
```
Comment (tanishqaggarwal) [inline test/test_quake/quake_common.h]: The control cycle length is 120, why use 140?

File: test/test_quake/quake_common.h [hunk 3]
```diff
@@ -34,3 +35,45 @@ void setUp(void) {
...
+    int retCode     = -1;   // return code/status
+    do
+    {
+        delay(DEFAULT_CTRL_CYCLE_LENGTH);
+        retCode = (q.*callback)();
+        ++numCycles;
+    }
+    while (retCode == Devices::PORT_UNAVAILABLE);
+    
+    Serial.printf("Number of cycles before %s: %d\r\n", fnName, numCycles);
```
Comment (tanishqaggarwal) [inline test/test_quake/quake_common.h]: Shouldn't we make an assertion about the number of cycles? I thought it should be at most 1

File: test/test_quake/test_quake.cpp [hunk 4]
```diff
@@ -2,91 +2,83 @@
 #include <string>
```
Comment (tanishqaggarwal) [inline test/test_quake/test_quake.cpp]: Super clean test file. Noice.

File: test/test_quake_controller/test_quake_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  task.set_downlink_msg("Blah", 4);
+  TEST_ASSERT_EQUAL(true, task.request_state(SBDWB));
+  exec_check(task, SBDWB, 1);
+
+  TEST_ASSERT_EQUAL(true, task.request_state(CONFIG));
+  TEST_ASSERT_EQUAL(CONFIG, task.get_current_state());
+  exec_check(task, CONFIG, 1);
+}
+
+  // Check that config can interrupt any state
```
Comment (tanishqaggarwal) [inline test/test_quake_controller/test_quake_controller.cpp]: Check that "idle" can interrupt

File: test/test_quake_controller/test_quake_controller.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+void test_config()
+{
+  StateFieldRegistry registry;
+  QuakeControlTask task(registry, 0);
+  TEST_ASSERT_EQUAL(true, task.request_state(CONFIG));
+  exec_check(task, CONFIG, 1); 
+  exec_check(task, CONFIG, 2); 
+  exec_check(task, CONFIG, 3); 
+  exec_check(task, IDLE, 0);
+}
```
Comment (tanishqaggarwal) [inline test/test_quake_controller/test_quake_controller.cpp]: Can you add one-line docstrings for these functions as well?

File: test/test_quake_controller/test_quake_controller.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+// Check that changing state is not ok and does not change the fn number or state
+void test_sbdwb_noint()
+{
+  StateFieldRegistry registry;
+  QuakeControlTask task(registry, 0);
+  TEST_ASSERT_EQUAL(true, task.request_state(CONFIG));
+  exec_check(task, CONFIG, 1);
+  TEST_ASSERT_EQUAL(CONFIG, task.get_current_state());
+}
```
Comment (tanishqaggarwal) [inline test/test_quake_controller/test_quake_controller.cpp]: How is this is testing changing state? The comment and/or the function name is wrong

File: test/test_quake_network/test_qlocate_network.cpp [hunk 8]
```diff
@@ -6,119 +6,62 @@
...
-        ;
-    delay(100);
-    while (!Serial3.available())
-        ;
-    // End SBD session
-    TEST_ASSERT_EQUAL(0, q.get_sbdix());
-    // Get SBDI response
-    const int *_pRes = q.get_sbdix_response();
+    const int *_pRes = q.sbdix_r;
     sbdix_r_t *pRes = (sbdix_r_t *)(_pRes);
```
Comment (tanishqaggarwal) [inline test/test_quake_network/test_qlocate_network.cpp]: If you want, you could replace this with the more intuitive functions available in QCT, right?

File: test/test_quake_manager/test_quake_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+  tf.quake_manager->dbg_transition_radio_state(radio_mode_t::config);
+  // Then my cycles should be updated
+  assert_radio(static_cast<unsigned int>(radio_mode_t::config));
+  TEST_ASSERT_EQUAL(initCycles+1, tf.quake_manager->dbg_get_last_checkin());
+}
+
+void test_no_more_cycles()
+{
+  TestFixture tf(static_cast<unsigned int>(radio_mode_t::config), CONFIG);
+  // If there are no more cycles, then transition to wait or config
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: You didn't actually assert that the transition went to wait or config

File: test/test_quake_manager/test_quake_manager.cpp [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+  tf.step();
+  // Then the error flag should be set and transition to WAIT
+  assert_radio(radio_mode_t::wait);
+  assert_qct(IDLE);
+  TEST_ASSERT_TRUE(tf.quake_manager->dbg_get_unexpectedFlag());
+}
+
+// ---------------------------------------------------------------------------
+// Helper functions
+// ---------------------------------------------------------------------------
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Are these helper functions? They look like pretty standalone tests

File: test/test_quake_manager/test_quake_manager.cpp [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+  TEST_ASSERT_EQUAL(70, tf.quake_manager->dbg_get_qct().dbg_get_MO_len());
+  check_buf_bytes(
+    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA", 
+    tf.quake_manager->dbg_get_qct().dbg_get_MO_msg(), tf.quake_manager->dbg_get_qct().dbg_get_MO_len()); 
+  tf.step();
+}
+
+void test_update_mo_same_snap()
+{
+  // If SBDWB has writtem the first piece of the snapshot
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: "Written"

File: test/test_quake_manager/test_quake_manager.cpp [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.radio_mo_packet_fp->get());
+  tf.step(); // sbdwb 0
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.quake_manager->dbg_get_qct().dbg_get_MO_msg());
+  tf.radio_mo_packet_fp->set(snap2); 
+  tf.step(); // 1
+  tf.quake_manager->dbg_get_qct().dbg_get_quake().sbdix_r[2] = 0; // have comms but no msg
+  tf.step(); // 2
+  tf.radio_mo_packet_fp->set(snap2); 
+  // Execute SBDIX
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.quake_manager->dbg_get_qct().dbg_get_MO_msg());
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Confused by:
- How SBDIX is getting executed
- Why you're setting the MO packet pointer to snap2 twice

File: test/test_quake_manager/test_quake_manager.cpp [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+          "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB"\
+          "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC"\
+          "DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD"\
+          "EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE";
+char* snap2 = (char*)
+          "1111111111111111111111111111111111111111111111111111111111111111111111"\
+          "2222222222222222222222222222222222222222222222222222222222222222222222"\
+          "3333333333333333333333333333333333333333333333333333333333333333333333"\
+          "4444444444444444444444444444444444444444444444444444444444444444444444"\
+          "555555555555555555555555555555555555555555555555555555555555555555555";
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Can you move these snapshot buffers down closer to the functions that actually use them in the test?

File: test/test_quake_manager/test_quake_manager.cpp [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+  for (int i = 0; i < size; i++)
+  {
+    if (buf1[i] != buf2[i])
+    {
+        TEST_ASSERT_EQUAL(buf1[i], buf2[i]);
+    }
+  }
+}
+
+void test_write_load_message()
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Need a comment above function

File: test/test_quake_manager/test_quake_manager.cpp [hunk 15]
```diff
@@ -0,0 +1,540 @@
...
+  tf.step(); // 1
+  tf.quake_manager->dbg_get_qct().dbg_get_quake().sbdix_r[2] = 0; // have comms but no msg
+  tf.step(); // 2
+  tf.radio_mo_packet_fp->set(snap2); 
+  // Execute SBDIX
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.quake_manager->dbg_get_qct().dbg_get_MO_msg());
+  tf.step(); // sbdix 0
+  tf.radio_mo_packet_fp->set(snap2); 
+  tf.step(); // sbdix 1
+  tf.radio_mo_packet_fp->set(snap2);
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Same here, why set the pointer twice?

File: test/test_quake_manager/test_quake_manager.cpp [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.quake_manager->dbg_get_qct().dbg_get_MO_msg());
+  // Return to SBDWB
+  tf.step(); // sbdwb 0
+  // Then the second piece of the snapshot should be loaded on the next write
+  check_buf_bytes(
+    "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB", 
+    tf.quake_manager->dbg_get_qct().dbg_get_MO_msg(), tf.quake_manager->dbg_get_qct().dbg_get_MO_len()); 
+}
+
+void test_update_mo_reset_idx()
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Comment before function

File: test/test_quake_manager/test_quake_manager.cpp [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+    "EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE", 
+    tf.quake_manager->dbg_get_qct().dbg_get_MO_msg(), tf.quake_manager->dbg_get_qct().dbg_get_MO_len()); 
+  tf.realSteps(2); // 1, 2
+  tf.radio_mo_packet_fp->set(snap2);
+  // Then mo_idx should be reset to 0 but mo_buffer should still be the same
+  TEST_ASSERT_EQUAL(0, tf.quake_manager->dbg_get_mo_idx());
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.quake_manager->dbg_get_mo_buffer_copy());
+}
+
+void test_update_mo_load_new_snap()
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Comment before function

File: test/test_quake_manager/test_quake_manager.cpp [hunk 18]
```diff
@@ -0,0 +1,540 @@
...
+{
+  // If have written the last piece of the first snapshot
+  TestFixture tf(static_cast<unsigned int>(radio_mode_t::write), SBDWB);
+  // Read Snap 1
```
Comment (tanishqaggarwal) [inline test/test_quake_manager/test_quake_manager.cpp]: Maybe point out here that snap1 is already loaded into the MO buffer, in the test fixture.

File: test/test_quake_manager/test_quake_manager.cpp [hunk 19]
```diff
@@ -0,0 +1,534 @@
...
+    assert_radio(radio_mode_t::config);
+    assert_qct(CONFIG);
+    assert_fn_num(0);
+    TEST_ASSERT_EQUAL(0, tf.quake_manager->mo_idx);
+    TEST_ASSERT_FALSE(tf.quake_manager->unexpectedFlag);
+}
+
+void test_dispatch_manual()
+{
+  TestFixture tf(static_cast<unsigned int>(radio_mode_t::manual), IDLE);
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply test/test_quake_manager/test_quake_manager.cpp]: Ah, I see that this test is commented out below--why?

File: src/FCCode/QuakeManager.cpp [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+{ 
+    // Retrieve fields from registry
+    control_cycle_count_fp = find_readable_field<unsigned int>("pan.cycle_no", __FILE__, __LINE__);
+    snapshot_size_fp = find_internal_field<unsigned int>("downlink_producer.snap_size", __FILE__, __LINE__);
+    radio_mo_packet_fp = find_internal_field<char*>("downlink_producer.mo_ptr", __FILE__, __LINE__);
+    radio_mt_packet_fp = find_internal_field<char*>("downlink_producer.mt_ptr", __FILE__, __LINE__);
+    radio_err_fp = find_internal_field<int>("downlink_producer.radio_err_ptr", __FILE__, __LINE__);
+
+    // Initialize Quake Manager variables
+    last_checkin_cycle = control_cycle_count_fp->get();
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/FCCode/QuakeManager.cpp]: I see, this is an artifact of `initCycles`. But why can't `initCycles` be zero?

File: src/FCCode/QuakeManager.cpp [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+
+    int err_code = qct.execute();
+
+    // If we are done with SBDRB --> save message and load next message
+    if (qct.get_current_state() == IDLE)
+    {
+        printf(debug_severity::info, 
+            "[Quake Info] SBDRB finished, transitioning to SBDWB\r\n");
+
+        radio_mt_packet_fp->set(qct.get_MT_msg());
```
Comment (tanishqaggarwal) [inline src/FCCode/QuakeManager.cpp]: Since get_MT_msg never changes its address, we can just do this assignment of radio_mt_packet_fp inside the constructor.

As we discussed, maybe you can replace this with a validity flag assignment instead.

File: test/test_quake/test_quake.cpp [hunk 22]
```diff
@@ -2,91 +2,83 @@
...
-    delay(1000 * 30);
+
     // Second time should fail because still waiting
-    TEST_ASSERT_EQUAL(Devices::PORT_UNAVAILABLE, q.get_sbdwb());
     delay(1000 * 30);
+    TEST_ASSERT_EQUAL(Devices::PORT_UNAVAILABLE, q.get_sbdwb());
+
     // Third time should succeed
+    delay(1000 * 30);
     TEST_ASSERT_EQUAL(Devices::TIMEOUT, q.get_sbdwb());
```
Comment (shihaocao) [inline test/test_quake/test_quake.cpp]: Why does TIMEOUT mean success?

File: test/test_quake/test_quake.cpp [hunk 23]
```diff
@@ -2,91 +2,83 @@
...
-    // test that we can send the minimum number of bytes
-    delay(DEFAULT_DELAY);
-    TEST_ASSERT_EQUAL(Devices::OK, q.query_sbdwb_1(1));
+}
+
+void test_bad_checksum(void) {
     // test bad checksum by sending extraneous data
-    delay(DEFAULT_DELAY);
+    delay(DEFAULT_CTRL_CYCLE_LENGTH);
     TEST_ASSERT_EQUAL(0, q.query_sbdwb_2("ZZ", 2));
```
Comment (shihaocao) [inline test/test_quake/test_quake.cpp]: Why does this cause a bad checksum?

File: src/FCCode/QuakeManager.h [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * Local copy of max_snapshot_size given by DownlinkProducer
+     */
+    size_t max_snapshot_size;
+
+    /**
+     * Temporary buffer of size max_snapshot_size + 1
+     * Only SBDWB may write to mo_buffer_copy or mo_idx
+     */
+    char* mo_buffer_copy;
```
Comment (shihaocao) [inline src/FCCode/QuakeManager.h]: Why max_snapshot_size + 1?
Also: maybe reword to snap_shot_size +1?
Since max_snapshot_size is already snap_shot_size +1

File: lib/Drivers/QLocate.hpp [hunk 25]
```diff
@@ -38,42 +36,30 @@ static constexpr int WRITE_FAIL = -30;          // failed to send command (write
...
- * the driver state. Calling the next "query_" method too early will likely
- * return PORT_UNAVAILABLE. Calling these methods in the incorrect state
- * will return WRONG_STATE.
+ * Calling the next "query_" method too early will likely return 
+ * PORT_UNAVAILABLE. 
  *
  * "get_" methods end the command sequence of the operation and will set the
- * state of the driver to IDLE provided that it reads data from the port. If
- * the method returns PORT_UNAVAILABLE or TIMEOUT
- * then it will not reset state to IDLE
```
Comment (shihaocao) [inline lib/Drivers/QLocate.hpp]: wait why did you delete these lines? did you forget to add them back in?

File: test/test_quake_manager/test_quake_manager.cpp (hunk 13 above)
Comment (shihaocao) [inline test/test_quake_manager/test_quake_manager.cpp]: Why does the line of 5's only have 69 5's?

File: test/test_quake_manager/test_quake_manager.cpp [hunk 26]
```diff
@@ -0,0 +1,540 @@
...
+  tf.radio_mo_packet_fp->set(snap2); 
+  tf.step(); // sbdix 1
+  tf.radio_mo_packet_fp->set(snap2); 
+  TEST_ASSERT_EQUAL_STRING(snap1, tf.quake_manager->dbg_get_qct().dbg_get_MO_msg());
+  // Return to SBDWB
+  tf.step(); // sbdwb 0
+  // Then the second piece of the snapshot should be loaded on the next write
+  check_buf_bytes(
+    "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB", 
+    tf.quake_manager->dbg_get_qct().dbg_get_MO_msg(), tf.quake_manager->dbg_get_qct().dbg_get_MO_len());
```
Comment (shihaocao) [inline test/test_quake_manager/test_quake_manager.cpp]: To clarify, are we setting the mo packet to snap2, but asserting that it should continue to try finish dealing with snap1 first?
